[AQLProfile] Fix caching of agents with new available data

venkat1361 opened the pull request:
## Motivation
hsa_agent not provided by new api/rocprofiler-sdk and causes every device to have same id, in cases where gfxip is same but different CU config, pm4factory doesn't know the difference. This fix uses gfxip and CU count as a key for cache.

## Technical Details

https://github.com/ROCm/rocm-systems/blob/develop/projects/aqlprofile/src/core/pm4_factory.cpp#L54
https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp#L942

From rocprofiler-sdk we never get hsa_agent and it remains to be set with handle=0.
aqlprofile V1 API receives hsa_agent handle(https://github.com/ROCm/rocm-systems/blob/develop/projects/aqlprofile/src/core/aql_profile.cpp#L558). the newer API which rocprofiler-sdk doesn't take hsa_agent(https://github.com/ROCm/rocm-systems/blob/develop/projects/aqlprofile/src/core/counters.cpp#L322).

Alternate Solutions: 
- remove instances_t from Pm4Factory (not sure about performance impact.)
- add UUID(https://github.com/ROCm/rocm-systems/blob/develop/projects/aqlprofile/src/core/include/aqlprofile-sdk/aql_profile_v2.h#L165)

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

venky1306 committed:
> hsa_agent not provided by new api/rocprofiler-sdk and causes every device to have same id,
in cases where gfxip is same and config is different pm4factory doesn't know the difference. This fix uses gfxip and CU count as a key for cache.

jrmadsen commented on projects/aqlprofile/src/core/pm4_factory.h:
```diff
@@ -256,11 +256,15 @@ class Pm4Factory {
   struct instances_fncomp_t {
-    bool operator()(const hsa_agent_t& a, const hsa_agent_t& b) const {
-      return a.handle < b.handle;
+    bool operator()(const AgentInfo& a, const AgentInfo& b) const {
+      int cmp = strcmp(a.gfxip, b.gfxip);
+      if (cmp < 0) return true;
+      if (cmp > 0) return false;
+      // If gfxip strings are equal, compare cu_num
+      return a.cu_num < b.cu_num;
```
So HSA will have different handles for two GPUs with the same gfxip and the same cu_number. If you are getting rid of this uniqueness, you need to make it clear in the comments why it doesn't matter

venkat1361 replied:
in aqlprofile v2 code workflow we skipped hsa for hsa_agent info and rely on rocprofiler-sdk to provide info about agent. due to which all agents will have AgentInfo.dev_id.handle set to 0 and cache fails to differenciate btw gpus which causes issues where, node has different types of gpus.

Aqlprofile V1 calls hsa_iterate_agents which populates AgentInfo.dev_id.handle.

bing-ma replied:
This fix may not always work. I will have a discussion with Venkat to see how to properly fix this.

bwelton replied:
All of the information to uniquely identify an agent should be present in aqlprofile_agent_info_t. 

```
typedef struct {
  const char* agent_gfxip; /**< Agent GFXIP (HSA_AGENT_INFO_NAME or KFD.product_name) */
  uint32_t xcc_num;        /**< XCC's on the agent (HSA_AMD_AGENT_INFO_NUM_XCC or KFD.num_xcc) */
  uint32_t se_num;         /**< SE's on the agent (HSA_AMD_AGENT_INFO_NUM_SHADER_ENGINES or
                              KFD.num_shader_banks) */
  uint32_t cu_num; /**< CU's on the agent (HSA_AMD_AGENT_INFO_COMPUTE_UNIT_COUNT or KFD.cu_count) */
  uint32_t shader_arrays_per_se; /**< Shader arrays per SE of agent
                                    (HSA_AMD_AGENT_INFO_NUM_SHADER_ARRAYS_PER_SE or
                                    KFD.simd_arrays_per_engine)*/
} aqlprofile_agent_info_t;
```

This struct should (and needs to) contain all information that AQLProfile requires to construct packets for an agent. Rocprofiler-sdk registers agents with AQLProfile by filling in this struct and supplying it to AQLProfile. A comparison on this struct should be enough to determine if an agent is unique in AQLProfile terms (i.e. unique in the ways it cares about for constructing packets). 

If there are new things in the future that AQLProfile needs to know about an arch/agent to construct packets for it, they need to be in this struct. Those unique things do not (and should not) include HSA related types (i.e. hsa_agent_t).. 

As for why this registration system was introduced, this was done primarily to limit the requirement for AQLProfile to link to HSA/etc to gain information it needs to create packets. It also allows for testing packet generation on machines where the hardware does not exist.

bing-ma replied:
Hi @bwelton, all the info in the current structure is only about GC. There are SKUs with same GC core but different memory settings which will affect non-gc blocks such as UMC and DF. Since we already got device_id from sysfs (not from rocr), passing device_id to aqlprofile will at least give aqlprofile an extra info to differentiate these different GPU SKUs.

bwelton replied:
>  passing device_id to aqlprofile will at least give aqlprofile an extra info to differentiate these different GPU SKUs.

How does AQLProfile know what this device_id means in relation to the above memory settings? How does a human know that these settings are set for a specific device_id or what their impact is? 

Why not just include this information into the struct such that when a new SKU comes out, we don't need to make code changes to support it?

bing-ma replied:
Somewhere in the software/firmware stack these info will be hardcoded. I am totally OK for SDK to pass the entire sysfs info to aqlprofile so that aqlprofile doesn't need to check device_id, but now only GC info is passed down.

venkat1361 committed:
Change comparison from gfxip to name in instances_fncomp_t

Updated comparison in instances_fncomp_t to use 'name' for backward compatibility with rocprofv2.

bing-ma reviewed (approved):
Looks good to me.
